### PR TITLE
FIX: handle arbitrary iterables in setp

### LINKED ISSUE #6212: ENH? BUG? `pyplot.setp`/`Artist.setp` does not accept non-indexable iterables of handles.
Setup
-----
`matplotlib.__version__` = 1.5.1
`matplotlib.__version__numpy__` = 1.6
Python 3.5.1, IPython 4.1.2  running under anaconda
OS: Red Hat 6.5

Matplotlib installed with `conda install matplotlib`.

Issue
-----
Supplying an iterable that is not a list or a tuple (or other indexable sequence) causes an issue with `setp`. This is probably OK with Python 2, but it is counter-intuitive for Python 3. I would expect the following to work, but it does not:

    >>> import matplotlib.pyplot as plt
    >>> import itertools
    >>> lines1 = plt.plot(range(3), range(3), range(5), range(5))
    >>> lines2 = plt.plot(range(4), range(4), range(6), range(6))
    >>> plt.setp(itertools.chain(lines1, lines2), color='red')
    Traceback (most recent call last):

      File "<ipython-input-6-2f274dd0d4c1>", line 1, in <module>
        plt.setp(itertools.chain(lines1, lines2), color='red')

      File "/home/jfoxrabi/miniconda3/lib/python3.5/site-packages/matplotlib/pyplot.py", line 351, in setp
        return _setp(*args, **kwargs)

      File "/home/jfoxrabi/miniconda3/lib/python3.5/site-packages/matplotlib/artist.py", line 1437, in setp
        insp = ArtistInspector(obj)

      File "/home/jfoxrabi/miniconda3/lib/python3.5/site-packages/matplotlib/artist.py", line 1032, in __init__
        if cbook.iterable(o) and len(o):

    TypeError: object of type 'itertools.chain' has no len()

The expected behavior would be the same as for

    >>> plt.setp(lines1 + lines2, color='red')

This example is a bit contrived to illustrate the issue, but I have a case where I am doing stuff to a bunch of lines from different places, then trying to set a couple of properties on them. Wrapping the iterator I have in `list()` is not a huge problem, but it is unexpected and probably should not be needed.

The docs are trivial to update in this case (I have already started a PR), but I will need to trace through a couple of things before I am sure I am sticking `next(iter(x))` instead of `x[0]` in the right place.

### PR DESCRIPTION
**Pull request by tacaswell**
closes #6212

Alternative / complimentary to #6217

I had this half done when @madphysicist said they would work on it, fixes the same problem but in a different place.

### CODE REVIEW

**Comment by madphysicist** [comment]
"He" is fine BTW.

File: lib/matplotlib/artist.py [hunk 1]
```diff
@@ -1437,7 +1437,12 @@ def setp(obj, *args, **kwargs):
       >>> setp(lines, linewidth=2, color='r')        # python style
     """
 
-    insp = ArtistInspector(obj)
```
Comment (madphysicist) [inline lib/matplotlib/artist.py]: Is `ArtistInspector` used elsewhere? If not, this approach is better.

File: lib/matplotlib/tests/test_artist.py [hunk 2]
```diff
@@ -187,6 +189,12 @@ def test_properties():
         assert len(w) == 0
 
 
+@cleanup
```
Comment (madphysicist) [inline lib/matplotlib/tests/test_artist.py]: This test alone should be reason enough to stick with your PR. You can add the example I gave in my PR if you want something less contrived.

File: lib/matplotlib/artist.py (hunk 1 above)
Comment (tacaswell) [inline lib/matplotlib/artist.py]:
```
14:24 $ git grep ArtistInspector
CHANGELOG:           hardcopy and other use.  Ther ArtistInspector could use a
doc/users/whats_new.rst:    differently for hardcopy and other use.  The ArtistInspector
lib/matplotlib/artist.py:        return ArtistInspector(self).properties()
lib/matplotlib/artist.py:class ArtistInspector(object):
lib/matplotlib/artist.py:        the :class:`~matplotlib.artist.ArtistInspector`.
lib/matplotlib/artist.py:        insp = ArtistInspector(obj)
lib/matplotlib/artist.py:    insp = ArtistInspector(objs[0])
lib/matplotlib/artist.py:        return '\n'.join(ArtistInspector(a).pprint_setters_rest(
lib/matplotlib/artist.py:        return '\n'.join(ArtistInspector(a).pprint_setters(leadingspace=2))
```

It hopefully many of these use cases will be replaced by traitlet based methods.

Comment (madphysicist) [inline lib/matplotlib/artist.py]: I just saw the same thing with regular `grep -r`. Would you mind if I absorbed some of your changes into my commit in the meantime?

Comment (tacaswell) [inline lib/matplotlib/artist.py]: feel free to cherry pick this to your branch or just re-write it.

File: lib/matplotlib/tests/test_artist.py [hunk 3]
```diff
@@ -187,6 +189,12 @@ def test_properties():
...
+def test_setp():
+    lines1 = plt.plot(range(3))
+    lines2 = plt.plot(range(3))
+    martist.setp(chain(lines1, lines2), 'lw', 5)
+
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_artist.py]: + One more empty line.

**Comment by tacaswell** [comment]
Folded into   #6217

**Pull request closed without merging**